File: public/files/js/pages/query.ts
/*
 * Copyright (c) 2013 Charles University in Prague, Faculty of Arts,
 *                    Institute of the Czech National Corpus
 * Copyright (c) 2013 Tomas Machalek <tomas.machalek@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2
 * dated June, 1991.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { Dict, Ident, List, pipe, tuple } from 'cnc-tskit';

import * as Kontext from '../types/kontext';
import { PageModel } from '../app/page';
import { TextTypesModel } from '../models/textTypes/main';
import { FirstQueryFormModel } from '../models/query/first';
import { WithinBuilderModel } from '../models/query/withinBuilder';
import { VirtualKeyboardModel } from '../models/query/virtualKeyboard';
import { QueryContextModel } from '../models/query/context';
import { UsageTipsModel } from '../models/usageTips';
import { init as queryFormInit, QueryFormProps } from '../views/query/first';
import { init as basicOverviewViewsInit } from '../views/query/basicOverview';
import * as PluginInterfaces from '../types/plugins';
import { PluginName } from '../app/plugin';
import { KontextPage } from '../app/main';
import {
    ConcLinesStorage, StorageUsingState,
    openStorage } from '../models/concordance/selectionStorage';
import { Actions as GlobalActions } from '../models/common/actions';
import corplistComponent from 'plugins/corparch/init';
import liveAttributes from 'plugins/liveAttributes/init';
import tagHelperPlugin from 'plugins/taghelper/init';
import { QueryHelpModel } from '../models/help/queryHelp';
import { ConcFormArgs, QueryFormArgs } from '../models/query/formArgs';
import { QuickSubcorpModel } from '../models/subcorp/quickSubcorp';
import { importInitialTTData, TTInitialData } from '../models/textTypes/common';
import { ConcServerArgs } from '../models/concordance/common';
import { AnyTTSelection, ExportedSelection } from '../types/textTypes';
import { Root } from 'react-dom/client';


/**
 *
 */
export class QueryPage {

    private layoutModel:PageModel;

    private queryModel:FirstQueryFormModel;

    private textTypesModel:TextTypesModel;

    private quickSubcorpModel:QuickSubcorpModel;

    private liveAttrsPlugin:PluginInterfaces.LiveAttributes.IPlugin;

    private queryHintModel:UsageTipsModel;

    private queryHelpModel:QueryHelpModel;

    private withinBuilderModel:WithinBuilderModel;

    private virtualKeyboardModel:VirtualKeyboardModel;

    private queryContextModel:QueryContextModel;

    private queryFormRoot:Root;

    private topbarHelpRoot:Root;

    private queryOverviewRoot:Root;


    constructor(layoutModel:PageModel, clStorage:ConcLinesStorage<StorageUsingState>) {
        this.layoutModel = layoutModel;
    }

    getConf<T>(name:string):T {
        return this.layoutModel.getConf<T>(name);
    }

    translate(msg:string, values?:{[k:string]:string}):string {
        return this.layoutModel.translate(msg, values);
    }

    private initCorplistComponent(widgetId:string):[PluginInterfaces.Corparch.WidgetView, PluginInterfaces.Corparch.IPlugin] {
        const plg = corplistComponent(this.layoutModel.pluginApi());
        return tuple(
            plg.createWidget(
                widgetId,
                'query',
                (corpora:Array<string>, subcorpId:string) => {
                    this.layoutModel.dispatcher.dispatch<typeof GlobalActions.SwitchCorpus>({
                        name: GlobalActions.SwitchCorpus.name,
                        payload: {
                            corpora,
                            subcorpus: subcorpId
                        }
                    });
                }
            ),
            plg
        );
    }

    createTTViews(
        queryFormArgs:QueryFormArgs,
        textTypesData:TTInitialData,
        subcorpTTStructure:ExportedSelection
    ):[QueryFormProps, Array<AnyTTSelection>] {
        const attributes = importInitialTTData(
            textTypesData, subcorpTTStructure, subcorpTTStructure);
        this.textTypesModel = new TextTypesModel({
                dispatcher: this.layoutModel.dispatcher,
                pluginApi: this.layoutModel.pluginApi(),
                attributes,
                readonlyMode: false,
                bibIdAttr: textTypesData.bib_id_attr,
                bibLabelAttr: textTypesData.bib_label_attr
        });
        const hasSelectedItems = this.textTypesModel.applyCheckedItems(
            queryFormArgs.selected_text_types,
            queryFormArgs.bib_mapping
        );

        const availableAlignedCorpora = this.layoutModel.getConf<Array<Kontext.AttrItem>>('availableAlignedCorpora');
        this.liveAttrsPlugin = liveAttributes(
            this.layoutModel.pluginApi(),
            this.layoutModel.pluginTypeIsActive(PluginName.LIVE_ATTRIBUTES),
            false,
            {
                bibIdAttr: textTypesData.bib_id_attr,
                bibLabelAttr: textTypesData.bib_label_attr,
                availableAlignedCorpora,
                refineEnabled: hasSelectedItems,
                manualAlignCorporaMode: false,
                subcorpTTStructure,
                textTypesData: this.layoutModel.getConf<TTInitialData>('textTypesData')
            }
        );

        let liveAttrsViews:PluginInterfaces.LiveAttributes.Views;
        if (this.layoutModel.pluginTypeIsActive(PluginName.LIVE_ATTRIBUTES)) {
            liveAttrsViews = this.liveAttrsPlugin.getViews(null, this.textTypesModel, !List.empty(availableAlignedCorpora));
            this.textTypesModel.enableAutoCompleteSupport();

        } else {
            liveAttrsViews = {
                LiveAttrsCustomTT: null,
                LiveAttrsView: null
            };
        }

        this.quickSubcorpModel = new QuickSubcorpModel(
            this.layoutModel.dispatcher,
            this.layoutModel,
            this.liveAttrsPlugin.isActive()
        );

        return tuple(
            {
                ...liveAttrsViews,
                formType: Kontext.ConcFormTypes.QUERY,
                tagHelperViews: {},
                allowCorpusSelection: null
            },
            attributes
        );
    }

    private initQueryModel(
        thPlugin:PluginInterfaces.TagHelper.IPlugin,
        queryFormArgs:QueryFormArgs,
        initialTTSelection:Array<AnyTTSelection>,
        bibIdAttr:string,
        bibLabelAttr:string
    ):void {
        const corpora = [this.layoutModel.getCorpusIdent().id].concat(
            this.layoutModel.getConf<Array<string>>('alignedCorpora') || []);
        this.queryModel = new FirstQueryFormModel({
            dispatcher: this.layoutModel.dispatcher,
            pageModel: this.layoutModel,
            quickSubcorpActive: Dict.size(TextTypesModel.exportSelections(
                initialTTSelection,
                bibIdAttr,
                bibLabelAttr,
                false,
                true,
            )) > 0,
            queryContextModel: this.queryContextModel,
            qsPlugin: this.layoutModel.qsuggPlugin,
            thPlugin,
            props: {
                corpora,
                availableAlignedCorpora: this.layoutModel.getConf<Array<Kontext.AttrItem>>(
                    'availableAlignedCorpora'
                ),
                currQueryTypes: queryFormArgs.curr_query_types,
                currQueries: queryFormArgs.curr_queries,
                currParsedQueries: queryFormArgs.curr_parsed_queries,
                currPcqPosNegValues: queryFormArgs.curr_pcq_pos_neg_values,
                currIncludeEmptyValues: queryFormArgs.curr_include_empty_values,
                currLposValues: queryFormArgs.curr_lpos_values,
                currQmcaseValues: queryFormArgs.curr_qmcase_values,
                currDefaultAttrValues: queryFormArgs.curr_default_attr_values,
                currUseRegexpValues: queryFormArgs.curr_use_regexp_values,
                subcorpList: this.layoutModel.getConf<Array<Kontext.SubcorpListItem>>(
                    'SubcorpList'
                ),
                currentSubcorp: this.layoutModel.getCorpusIdent().usesubcorp,
                subcorpusId: this.layoutModel.getCorpusIdent().usesubcorp,
                isForeignSubcorpus: this.layoutModel.getCorpusIdent().foreignSubcorp,
                tagsets: queryFormArgs.tagsets,
                shuffleConcByDefault: this.layoutModel.getConf<boolean>('ShuffleConcByDefault'),
                forcedAttr: this.layoutModel.getConf<string>('ForcedAttr'),
                attrList: this.layoutModel.getConf<Array<Kontext.AttrItem>>('AttrList'),
                structAttrList: Kontext.structsAndAttrsToStructAttrList(this.layoutModel.getConf<Kontext.StructsAndAttrs>('structsAndAttrs')),
                structList: Kontext.structsAndAttrsToStructList(this.layoutModel.getConf<Kontext.StructsAndAttrs>('structsAndAttrs')),
                wPoSList: this.layoutModel.getConf<Array<{v:string; n:string}>>('Wposlist'),
                inputLanguages: this.layoutModel.getConf<{[corpname:string]:string}>(
                    'InputLanguages'
                ),
                textTypesNotes: this.layoutModel.getConf<string>('TextTypesNotes'),
                bibIdAttr,
                selectedTextTypes: queryFormArgs.selected_text_types,
                hasLemma: queryFormArgs.has_lemma,
                useRichQueryEditor:this.layoutModel.getConf<boolean>('UseRichQueryEditor'),
                isAnonymousUser: this.layoutModel.getConf<boolean>('anonymousUser'),
                isLocalUiLang: this.layoutModel.getConf<boolean>('isLocalUiLang'),
                suggestionsConfigured: this.layoutModel.getConf<boolean>('QSEnabled'),
                simpleQueryDefaultAttrs: pipe(
                    [...this.layoutModel.getConf<Array<string>>('alignedCorpora')],
                    List.push(this.layoutModel.getCorpusIdent().id),
                    List.map(corp => tuple(
                        corp,
                        pipe(
                            this.layoutModel.getConf<Array<Kontext.AttrItem>>('AttrList'),
                            List.map(v => v.n),
                            this.layoutModel.getConf<Array<string>>('SimpleQueryDefaultAttrs').length > 0 ?
                                List.unshift<string|Array<string>>(this.layoutModel.getConf<Array<string>>('SimpleQueryDefaultAttrs')) :
                                v => v
                        )
                    )),
                    Dict.fromEntries()
                ),
                concViewPosAttrs: this.getConf<ConcServerArgs>('currentArgs').attrs,
                alignCommonPosAttrs: this.getConf<Array<string>>('AlignCommonPosAttrs'),
                concPreflight: this.getConf<Kontext.PreflightConf|null>('concPreflight')
            }
        });
    }


    private attachQueryForm(
        properties:QueryFormProps,
        corparchWidget:PluginInterfaces.Corparch.WidgetView,
        corparchWidgetId:string

    ):void {
        const queryFormComponents = queryFormInit({
            dispatcher: this.layoutModel.dispatcher,
            he: this.layoutModel.getComponentHelpers(),
            CorparchWidget: corparchWidget,
            corparchWidgetId,
            queryModel: this.queryModel,
            textTypesModel: this.textTypesModel,
            quickSubcorpModel: this.quickSubcorpModel,
            queryHintModel: this.queryHintModel,
            withinBuilderModel: this.withinBuilderModel,
            virtualKeyboardModel: this.virtualKeyboardModel,
            queryContextModel: this.queryContextModel,
            querySuggest: this.layoutModel.qsuggPlugin,
            queryHelpModel: this.queryHelpModel,
            searchHistoryModel: this.layoutModel.getModels().searchHistoryModel
        });
        this.queryFormRoot = this.layoutModel.renderReactComponent(
            queryFormComponents.QueryForm,
            window.document.getElementById('query-form-mount'),
            properties
        );
        this.topbarHelpRoot = this.layoutModel.renderReactComponent(
            queryFormComponents.QueryHelp,
            window.document.getElementById('topbar-help-mount'),
            {
                isLocalUiLang: this.layoutModel.getConf<boolean>('isLocalUiLang')
            }
        );
    }

    private initCorpnameLink():void {
        const queryOverviewViews = basicOverviewViewsInit(
            this.layoutModel.dispatcher,
            this.layoutModel.getComponentHelpers(),
            this.layoutModel.getModels().mainMenuModel
        );
        this.queryOverviewRoot = this.layoutModel.renderReactComponent(
            queryOverviewViews.EmptyQueryOverviewBar,
            window.document.getElementById('query-overview-mount'),
        );
    }

    init():void {
        this.layoutModel.init(true, [], () => {
            this.queryHintModel = new UsageTipsModel(
                this.layoutModel.dispatcher,
                this.layoutModel.translate.bind(this.layoutModel)
            );
            this.withinBuilderModel = new WithinBuilderModel(
                this.layoutModel.dispatcher,
                this.layoutModel
            );

            this.virtualKeyboardModel = new VirtualKeyboardModel(
                this.layoutModel.dispatcher,
                this.layoutModel
            );

            const concFormsArgs = this.layoutModel.getConf<{[ident:string]:ConcFormArgs}>(
                'ConcFormsArgs'
            );

            let queryId = '__new__';
            const q = this.layoutModel.getNestedConf<string>('currentArgs', 'q');
            if (q.length > 0) {
                queryId = q[0].replace('~', '');
            }

            const queryFormArgs = concFormsArgs[queryId] as QueryFormArgs;
            this.queryContextModel = new QueryContextModel(
                this.layoutModel.dispatcher,
                queryFormArgs
            );

            this.queryHelpModel = new QueryHelpModel(
                this.layoutModel,
                this.layoutModel.dispatcher,
                {
                    isBusy: false,
                    rawHtml: '',
                    tagsets: queryFormArgs.tagsets,
                    activeCorpora: [
                        this.layoutModel.getCorpusIdent().id,
                        ...this.layoutModel.getConf<Array<string>>('alignedCorpora')
                    ]
                }
            );

            const textTypesData = this.layoutModel.getConf<TTInitialData>('textTypesData');
            const subcorpTTStructure = this.layoutModel.getConf<ExportedSelection>('SubcorpTTStructure');
            const [ttAns, ttSelection] = this.createTTViews(queryFormArgs, textTypesData, subcorpTTStructure);

            const tagBuilderCorpora = [
                this.layoutModel.getCorpusIdent().id,
                ...List.map(
                    v => v.n,
                    this.layoutModel.getConf<Array<Kontext.AttrItem>>('availableAlignedCorpora')
                )
            ];
            const tagHelperPlg = tagHelperPlugin(this.layoutModel.pluginApi());
            ttAns.tagHelperViews = pipe(
                this.layoutModel.isNotEmptyPlugin(tagHelperPlg) ? tagBuilderCorpora : [],
                List.map(corpus => tuple(
                        corpus,
                        tagHelperPlg.getWidgetView(
                            corpus,
                            corpus,
                            this.layoutModel.getNestedConf<
                            Array<PluginInterfaces.TagHelper.TagsetInfo>>(
                                'pluginData', 'taghelper', 'corp_tagsets')
                        )
                )),
                Dict.fromEntries()
            );

            ttAns.allowCorpusSelection = true;

            this.initQueryModel(
                tagHelperPlg,
                queryFormArgs,
                ttSelection,
                textTypesData.bib_id_attr,
                textTypesData.bib_label_attr
            );
            const corparchWidgetId = Ident.puid()
            const [corparchWidget, corparchPlg]  = this.initCorplistComponent(corparchWidgetId);
            this.attachQueryForm(
                ttAns,
                corparchWidget,
                corparchWidgetId
            );
            this.initCorpnameLink();
            // all the models must be unregistered and components must
            // be unmounted to prevent memory leaks and unwanted action handlers
            // from previous instance
            this.layoutModel.registerCorpusSwitchAwareModels(
                () => {
                    this.layoutModel.unmountReactComponent(this.queryFormRoot);
                    this.layoutModel.unmountReactComponent(this.queryOverviewRoot);
                    this.layoutModel.unmountReactComponent(this.topbarHelpRoot);
                    this.init();
                },
                this.queryModel,
                corparchPlg,
                this.queryHintModel,
                this.textTypesModel,
                this.queryContextModel,
                this.withinBuilderModel,
                this.virtualKeyboardModel,
                this.liveAttrsPlugin,
                this.quickSubcorpModel
            );
        });
    }
}


export function init(conf:Kontext.Conf):void {
    const layoutModel = new KontextPage(conf);
    const clStorage:ConcLinesStorage<StorageUsingState> = openStorage(
        layoutModel.dispatcher,
        (err:Error) => {
            layoutModel.showMessage('error', err);
        }
    );
    const pageModel = new QueryPage(layoutModel, clStorage);
    pageModel.init();
}

File: lib/action/argmapping/conc/query.py
# Copyright (c) 2017 Charles University, Faculty of Arts,
#                    Institute of the Czech National Corpus
# Copyright (c) 2017 Tomas Machalek <tomas.machalek@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 2
# dated June, 1991.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

from typing import Dict, Any, List, Tuple
from dataclasses import dataclass, field
from dataclasses_json import dataclass_json

import plugins
from action.plugin.ctx import PluginCtx
from action.argmapping.error import ArgumentMappingError
from action.argmapping.conc.base import ConcFormArgs


@dataclass_json
@dataclass
class _QueryFormArgs:
    form_type: str = 'query'
    curr_query_types: Dict[str, str] = field(default_factory=dict)
    curr_queries: Dict[str, str] = field(default_factory=dict)
    curr_parsed_queries: Dict[str, str] = field(default_factory=dict)
    curr_pcq_pos_neg_values: Dict[str, str] = field(default_factory=dict)
    curr_include_empty_values: Dict[str, bool] = field(default_factory=dict)
    curr_lpos_values: Dict[str, str] = field(default_factory=dict)
    curr_qmcase_values: Dict[str, bool] = field(default_factory=dict)
    curr_default_attr_values: Dict[str, bool] = field(default_factory=dict)
    curr_use_regexp_values: Dict[str, bool] = field(default_factory=dict)
    tagsets: Dict[str, List[Dict[str, Any]]] = field(default_factory=dict)
    has_lemma: Dict[str, bool] = field(default_factory=dict)
    asnc: bool = False
    no_query_history: bool = False

    # context filter
    fc_lemword_type: str = 'all'
    fc_lemword_wsize: Tuple[int, int] = (-5, 5)
    fc_lemword: str = ''
    fc_pos_type: str = 'all'
    fc_pos_wsize: Tuple[int, int] = (-5, 5)
    fc_pos: List[str] = field(default_factory=list)   # TODO is the type correct ??

    selected_text_types: Dict[str, List[str]] = field(default_factory=dict)
    # for bibliography structattr - maps from hidden ids to visible titles (this is optional)
    bib_mapping: Dict[str, str] = field(default_factory=dict)


def _corp_mapping(corpora: List[str], v: Any = None) -> Dict[str, Any]:
    return {c: v for c in corpora}


class QueryFormArgs(ConcFormArgs[_QueryFormArgs]):
    """
    QueryFormArgs provides methods to handle concordance
    query form arguments represented by the _QueryFormArgs data class.
    """

    def __init__(self, plugin_ctx: PluginCtx, corpora: List[str], persist: bool) -> None:
        super().__init__(persist)
        self._plugin_ctx = plugin_ctx
        self.data = _QueryFormArgs(
            form_type='query',
            curr_query_types=_corp_mapping(corpora, 'simple'),
            curr_queries=_corp_mapping(corpora),
            curr_parsed_queries=_corp_mapping(corpora),
            curr_pcq_pos_neg_values=_corp_mapping(corpora),
            curr_include_empty_values=_corp_mapping(corpora),
            curr_lpos_values=_corp_mapping(corpora),
            curr_qmcase_values=_corp_mapping(corpora),
            curr_default_attr_values=_corp_mapping(corpora),
            curr_use_regexp_values=_corp_mapping(corpora, False),
            tagsets=_corp_mapping(corpora),
            has_lemma=_corp_mapping(corpora))
        for corp in corpora:
            self._add_corpus_metadata(corp)

    def apply_last_used_opts(
            self, data: Dict[str, Any], prev_corpora: List[str], curr_corpora: List[str], curr_posattrs: List[str]):
        self._test_data_type(data, 'form_type', 'query')
        prev_maincorp = prev_corpora[0]
        curr_maincorp = curr_corpora[0]
        self.data.curr_query_types[curr_maincorp] = data['curr_query_types'][prev_maincorp]
        self.data.curr_qmcase_values[curr_maincorp] = data['curr_qmcase_values'][prev_maincorp]
        self.data.curr_use_regexp_values[curr_maincorp] = data['curr_use_regexp_values'][prev_maincorp]
        prev_default_attr = data['curr_default_attr_values'][prev_maincorp]
        if prev_default_attr in curr_posattrs:
            self.data.curr_default_attr_values[curr_maincorp] = prev_default_attr
        else:
            self.data.curr_default_attr_values[curr_maincorp] = None

    @staticmethod
    def _test_data_type(data, type_key, type_id):
        data_type = data.get(type_key)
        if data_type != type_id:
            raise ArgumentMappingError(
                f'Invalid form data type "{data_type}" for a query form mapping.')

    def update_by_user_query(self, data, bib_mapping):
        self._test_data_type(data, 'type', 'concQueryArgs')
        self.data.asnc = data.get('async', False)
        self.data.no_query_history = data.get('no_query_history', False)
        for query in data['queries']:
            corp = query['corpname']
            self.data.curr_query_types[corp] = query['qtype']
            self.data.curr_queries[corp] = query.get('query')
            self.data.curr_parsed_queries[corp] = query.get('queryParsed', [])
            self.data.curr_pcq_pos_neg_values[corp] = query['pcq_pos_neg']
            self.data.curr_include_empty_values[corp] = query['include_empty']
            self.data.curr_qmcase_values[corp] = query.get('qmcase', False)
            self.data.curr_default_attr_values[corp] = query.get('default_attr')
            self.data.curr_use_regexp_values[corp] = query.get('use_regexp', False)
        self.data.bib_mapping = bib_mapping

        ctx = data['context']
        self.data.fc_lemword_type = ctx.get('fc_lemword_type', self.data.fc_lemword_type)
        self.data.fc_lemword_wsize = ctx.get('fc_lemword_wsize', self.data.fc_lemword_wsize)
        self.data.fc_lemword = ctx.get('fc_lemword', self.data.fc_lemword)
        self.data.fc_pos_type = ctx.get('fc_pos_type', self.data.fc_pos_type)
        self.data.fc_pos_wsize = ctx.get('fc_pos_wsize', self.data.fc_pos_wsize)
        self.data.fc_pos = ctx.get('fc_pos', self.data.fc_pos)
        self.data.selected_text_types = data['text_types']

    def _add_corpus_metadata(self, corpus_id: str):
        with plugins.runtime.CORPARCH as ca, plugins.runtime.TAGHELPER as th:
            corp_info = ca.get_corpus_info(self._plugin_ctx, corpus_id)
            self.data.has_lemma[corpus_id] = corp_info.manatee.has_lemma
            self.data.tagsets[corpus_id] = [d.to_dict() for d in corp_info.tagsets]
            for ts in self.data.tagsets[corpus_id]:
                ts['widgetEnabled'] = ts['widgetEnabled'] and th.tags_available_for(
                    self._plugin_ctx, corpus_id, ts['ident'])

    def serialize(self) -> Dict[str, Any]:
        ans = super().to_dict()
        del ans['has_lemma']
        del ans['tagsets']
        return ans
